### PR TITLE
Translate transport, cross_surface, and particle restarts to C++

### PR DESCRIPTION
**Pull request by paulromano**
This PR moves the subroutines from tracking.F90 to be methods of the `Particle` class in C++. I've also moved particle restarts over as well, and together this enabled me to remove a bunch of Fortran files. One or two more of these PRs and I think we'll be done!

### CODE REVIEW

File: src/particle.cpp [hunk 1]
```diff
@@ -119,6 +132,506 @@ Particle::from_source(const Bank* src)
...
+
+      // Score flux derivative accumulators for differential tallies.
+      if (!model::active_tallies.empty()) score_collision_derivative(this);
+    }
+
+    // If particle has too many events, display warning and kill it
+    ++n_event;
+    if (n_event == MAX_EVENTS) {
+      if (mpi::master) warning("Particle " + std::to_string(id)
+        + " underwent maximum number of events.");
```
Comment (smharper) [inline src/particle.cpp]: Why is this only printed from the master process?

File: src/particle.cpp [hunk 2]
```diff
@@ -119,6 +132,506 @@ Particle::from_source(const Bank* src)
...
+      Position r {coord[0].xyz};
+      coord[0].xyz[0] -= TINY_BIT * coord[0].uvw[0];
+      coord[0].xyz[1] -= TINY_BIT * coord[0].uvw[1];
+      coord[0].xyz[2] -= TINY_BIT * coord[0].uvw[2];
+      score_surface_tally(this, model::active_meshsurf_tallies);
+      std::copy(&r.x, &r.x + 3, coord[0].xyz);
+    }
+
+    // Get a pointer to the partner periodic surface.  Offset the index to
+    // correct for C vs. Fortran indexing.
```
Comment (smharper) [inline src/particle.cpp]: Remove the bit about `Offset the index...`

File: src/particle.cpp [hunk 3]
```diff
@@ -119,6 +132,506 @@ Particle::from_source(const Bank* src)
...
+    auto surf_p = dynamic_cast<PeriodicSurface*>(surf);
+    if (surf_p) {
```
Comment (smharper) [inline src/particle.cpp]: Why the `if (surf_p)` check?  It seems if the surface is not a `PeriodicSurface` then something has gone very wrong and this `if` will be no help anyway.

File: src/particle.cpp [hunk 4]
```diff
@@ -119,6 +132,506 @@ Particle::from_source(const Bank* src)
...
+    // a void in the model, and 2) the particle hit a surface at a tangent. If
+    // the particle is really traveling tangent to a surface, if we move it
+    // forward a tiny bit it should fix the problem.
+
+    n_coord = 1;
+    coord[0].xyz[0] += TINY_BIT * coord[0].uvw[0];
+    coord[0].xyz[1] += TINY_BIT * coord[0].uvw[1];
+    coord[0].xyz[2] += TINY_BIT * coord[0].uvw[2];
+
+    // Couldn't find next cell anywhere// This probably means there is an actual
```
Comment (smharper) [inline src/particle.cpp]: Next time I'd suggest `^\s\+!` for the REGEX :p

File: src/particle.cpp (hunk 3 above)
Comment (paulromano) [reply src/particle.cpp]: Fair point, I'll remove the check

File: src/particle.cpp [hunk 5]
```diff
@@ -119,6 +132,506 @@ Particle::from_source(const Bank* src)
...
+  if (settings::verbosity >= 9 || simulation::trace) {
+     write_message("Simulating Particle " + std::to_string(id));
+  }
+
+  // Initialize number of events to zero
+  int n_event = 0;
+
+  // Add paricle's starting weight to count for normalizing tallies later
+#pragma omp atomic
+  total_weight += wgt;
```
[the remark this answers is not in the document]
Comment (paulromano) [reply src/particle.cpp]: Yes, I'll move it there

File: src/particle.cpp (hunk 4 above)
Comment (paulromano) [reply src/particle.cpp]: I'm not that sophisticated ;)

**Pull request merged**